docs/designs/keras_callback.md

```diff
@@ -0,0 +1,208 @@
+```python
+class CallbackList():
+    def on_batch_end(self, batch, logs=None):
+        for callback in self.callback_list:
+            callback.on_batch_end(bach, logs)
+```
+For detail, we can view the [source code](https://github.com/tensorflow/tensorflow/blob/cf7fcf164c9846502b21cebb7d3d5ccf6cb626e8/tensorflow/python/keras/callbacks.py#L189-L196
+) of `CallbackList` in Tensorflow.
+
+So, we can also use CallbackList to wrap the callbacks in ElasticDL.
+```python
+from tensorflow.python.keras.callbacks import CallbackList
+callbacks = callbacks()
+callback_list = CallbackList(callbacks)
+callback_list.on_batch_end(batch)
+```
+
+
+### Call the Methods of Callback using ParameterServerStrategy
+
+Using ParameterServerStrategy in ElasticDL, we will launch instances as master, worker and parameter server. 
+The master controls the start and end of job and epoch, the worker processes each batch and the parameter server update model weights with batch gradients. So we need to call different methods on different instance according the call frequency. Due to there are many methods in `tf.keras.callbacks.Callback` and we don't have use case for each method, so we only design to support calling methods which the use cases listed in the motivation section need. And we will design to support other methods if there is a use case.
+
+
+### Call Methods of Callback Per Batch on Worker
+
+The methods should be called per batch in `tf.keras.callbacks.Callback` are
+```python
+on_(train|test|predict)_batch_begin(self, batch, logs=None)
+on_(train|test|predict)_batch_end(self, batch, logs=None)
+```
+
+"Case 1" needs to customize behavior by calling `on_predict_batch_end`. 
+```python
+def on_predict_batch_end(self, batch, logs=None):
+    """Call on the prediction end of each batch
+    Args:
+        batch: integer, index of batch in the current worker
+        logs: dict. Has keys batch, size and predictions representing
+            the current batch number in the worker,
+            the size of the batch,
+            the prediction outputs of the batch.
+    """
+```
+
+On the worker, we can call `on_predict_batch_end` after forward processing with batch features:
+```python
+def _process_minibatch(self, task_type, features, batch_index):
+    ...
+    if task_type == elasticdl_pb2.PREDICTION:
+        predictions = self.forward_process(features)
+        prediction_logs = {
+            "batch": batch_index,
+            "size": len(predictions)
+            "predictions": predictions
+        }
+        self.callbacks_list.on_predict_batch_end(
+            batch=batch_index, logs=prediction_logs
+        )
+```
+
+### Call Methods of Callback Per Updating Batch Gradients on PS
+
+"Case 2" and "Case 3" needs to customize behaviors After updating model weights using batch gradients on the PS. So, we add `on_update_batch_end` method, though it is not in `tf.keras.callbacks.Callback`.
+```python
+def on_update_batch_end(self, version, logs=None):
```

Comment:
Are these two methods `on_update_batch_start|end` exposed to users?

Reply:
We can expose the two methods to users. But I tend to we provide some premade callbacks class using the two methods like [Keras premade callback classes](https://www.tensorflow.org/api_docs/python/tf/keras/callbacks#classes).